libc/string_cpy.c

```diff
@@ -0,0 +1,1269 @@
+
+	/* Creating expected value of connected tests strings */
+	strcpy(expVal, testStr);
+	strcat(expVal, testStr);
+
+	/* Copying string in same space but couple addresses onto left to merge text into one combining two similar texts */
+	sprintf(&memStr[BIG_NUMB / 2], "%s", testStr);
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(testStr, &memStr[BIG_NUMB / 2], sizeof(testStr));
+	TEST_ASSERT_EQUAL_PTR(NULL, memccpy(&memStr[(BIG_NUMB / 2) - 11], &memStr[BIG_NUMB / 2], 'x', sizeof(testStr) - 1));
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(&memStr[(BIG_NUMB / 2) - 11], expVal, sizeof(testStr) * 2 - 2);
+
+	/* Adding another one iteration of test string */
+	strcat(expVal, testStr);
+
+	/* Same as before but move onto right */
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(testStr, &memStr[BIG_NUMB / 2], sizeof(testStr));
+	TEST_ASSERT_EQUAL_PTR(NULL, memccpy(&memStr[(BIG_NUMB / 2) + 11], &memStr[BIG_NUMB / 2], 'x', sizeof(testStr) - 1));
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(&memStr[(BIG_NUMB / 2) - 11], expVal, sizeof(testStr) * 3 - 3);
+
+#endif
+}
+
+
+TEST(string_memccpy, one_byte)
+{
+#ifdef __phoenix__
+	TEST_IGNORE();
+#endif
+#ifndef __phoenix__
+
+	char testArray[MAX_ASCII_SET] = { 0 },
+		 input[MAX_ASCII_SET] = { 1 };
+	int i;
+
+	TEST_ASSERT_NOT_EQUAL_INT(0, strcmp(input, testArray));
+
+	for (i = 0; i < MAX_ASCII_SET; i++) {
+
+		/* This loop will copy only one byte in a place where the loop iterator pointing this place */
+		TEST_ASSERT_EQUAL_PTR(&testArray[i] + 1, memccpy(&testArray[i], &input[i], input[i], 1));
+	}
+
+	/* Checking if all elements were correctly copied */
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(input, testArray, MAX_ASCII_SET);
+
+#endif
+}
+
+
+TEST(string_memccpy, nothing)
+{
+#ifdef __phoenix__
+	TEST_IGNORE();
+#endif
+#ifndef __phoenix__
+
+	char testArray[MAX_STR_LEN] = TEST_STR1,
+		 input[MAX_STR_LEN] = { 0 },
+		 strText[] = "Lorem ipsum \0 dolor";
+
+	/* Clearing array with support array filled with 0 */
+	TEST_ASSERT_EQUAL_PTR(NULL, memccpy(testArray, input, 'x', MAX_STR_LEN));
+	TEST_ASSERT_NOT_EQUAL_INT(0, strcmp(testArray, TEST_STR1));
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(input, testArray, MAX_STR_LEN);
+
+
+	/* Checking no sensitivity for NULL */
+	memset(input, 0, sizeof(input));
+	TEST_ASSERT_EQUAL_PTR(NULL, memccpy(input, strText, 'x', sizeof(strText)));
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(strText, input, sizeof(strText));
+
+#endif
+}
+
+
+TEST(string_memccpy, big)
+{
+#ifdef __phoenix__
+	TEST_IGNORE();
+#endif
+#ifndef __phoenix__
+
+	char buff[BIG_NUMB] = { 0 };
+
+	/* Checking capability of handling big blocks of data */
+	TEST_ASSERT_EQUAL_INT(1, big_string());
+	memccpy(buff, long_str, '\xFF', BIG_NUMB);
+	TEST_ASSERT_EQUAL_STRING(buff, long_str);
+	TEST_ASSERT_EQUAL_CHAR(buff[0], '\1');
+	TEST_ASSERT_EQUAL_CHAR(buff[MAX_ASCII_SET / 4], 'A');
+	TEST_ASSERT_EQUAL_CHAR(buff[MAX_ASCII_SET / 2], '\x81');
+
+	/* Offset big chunk of data */
+	TEST_ASSERT_NOT_NULL(memccpy(&buff[128], long_str, '\xFF', BIG_NUMB - MAX_ASCII_SET));
+	TEST_ASSERT_EQUAL_CHAR(buff[0], '\1');
```

Comment:
You could simply check the whole content of buff using 3x `TEST_EQUAL_CHAR_ARRAY` - first 128 chars {1..127}, next 128 chars {1..127} and the rest {0, 0, ....}